iroh-bytes/src/store/bao_file.rs

```diff
@@ -0,0 +1,1238 @@
+    pub fn is_complete(&self) -> bool {
+        matches!(
+            self.storage.read().unwrap().deref(),
+            BaoFileStorage::Complete(_)
+        )
+    }
+
+    /// An AsyncSliceReader for the data file.
+    ///
+    /// Caution: this is a reader for the unvalidated data file. Reading this
+    /// can produce data that does not match the hash.
+    pub fn data_reader(&self) -> DataReader {
+        DataReader(Some(self.clone()))
+    }
+
+    /// An AsyncSliceReader for the outboard file.
+    ///
+    /// The outboard file is used to validate the data file. It is not guaranteed
+    /// to be complete.
+    pub fn outboard_reader(&self) -> OutboardReader {
+        OutboardReader(Some(self.clone()))
+    }
+
+    /// The most precise known total size of the data file.
+    pub fn current_size(&self) -> io::Result<u64> {
+        match self.storage.read().unwrap().deref() {
+            BaoFileStorage::Complete(mem) => Ok(mem.data_size()),
+            BaoFileStorage::IncompleteMem(mem) => Ok(mem.current_size()),
+            BaoFileStorage::IncompleteFile(file) => file.current_size(),
+        }
+    }
+
+    /// The outboard for the file.
+    pub fn outboard(&self) -> io::Result<PreOrderOutboard<OutboardReader>> {
+        let root = self.hash.into();
+        let tree = BaoTree::new(ByteNum(self.current_size()?), IROH_BLOCK_SIZE);
+        let outboard = self.outboard_reader();
+        Ok(PreOrderOutboard {
+            root,
+            tree,
+            data: outboard,
+        })
+    }
+
+    /// The hash of the file.
+    pub fn hash(&self) -> Hash {
+        self.hash
+    }
+
+    /// Create a new writer from the handle.
+    pub fn writer(&self) -> BaoFileWriter {
+        BaoFileWriter(Some(self.clone()))
+    }
+
+    /// This is the synchronous impl for writing a batch.
+    fn write_batch(&self, size: u64, batch: &[BaoContentItem]) -> io::Result<HandleChange> {
+        let mut storage = self.storage.write().unwrap();
+        match storage.deref_mut() {
+            BaoFileStorage::IncompleteMem(mem) => {
+                // check if we need to switch to file mode, otherwise write to memory
+                if max_offset(batch) <= self.config.max_mem as u64 {
+                    mem.write_batch(size, batch)?;
+                    Ok(HandleChange::None)
+                } else {
+                    // create the paths. This allocates 3 pathbufs, so we do it
+                    // only when we need to.
+                    let paths = self.config.paths(&self.hash);
+                    // *first* switch to file mode, *then* write the batch.
+                    //
+                    // otherwise we might allocate a lot of memory if we get
+                    // a write at the end of a very large file.
+                    let mut file_batch = mem.persist(paths)?;
+                    // if let Some(cb) = self.config.on_file_create.as_ref() {
+                    //     cb(&self.hash)?;
```

Comment:
what about this?

Reply:
The place where the cb happens has moved further outside. This can be removed.